src/PooledArrays.jl

```diff
@@ -165,31 +186,120 @@ Base.size(pa::PooledArray) = size(pa.refs)
 Base.length(pa::PooledArray) = length(pa.refs)
 Base.lastindex(pa::PooledArray) = lastindex(pa.refs)
 
-Base.copy(pa::PooledArray) = PooledArray(RefArray(copy(pa.refs)), copy(pa.invpool))
-# TODO: Implement copy_to()
+Base.copy(pa::PooledArray) = PooledArray(pa)
+
+Base.copyto!(dest::PooledArray{T, R, N, RA},
+             src::PooledArray{T, R, N, RA}) where {T, R, N, RA} =
+    copyto!(dest, 1, src, 1, length(src))
+
+# TODO: handle case when dest is SubArray
+
+Base.copyto!(dest::PooledArray{T, R, N, RA},
+             src::SubArray{T, N, PooledArray{T, R, M, RA}}) where {T, R, N, M, RA} =
+    copyto!(dest, 1, src, 1, length(src))
+
+function Base.copy!(dest::PooledArray{T, R, N, RA},
+                    src::PooledArray{T, R, N, RA}) where {T, R, N, RA}
+    copy!(dest.refs, src.refs)
+    Threads.atomic_add!(src.refcount, 1)
+    dest.pool = src.pool
+    dest.invpool = src.invpool
+    dest.refcount = src.refcount
+    return dest
+end
+
+function Base.copy!(dest::PooledArray{T, R, N, RA},
+                    src::SubArray{T, N, PooledArray{T, R, M, RA}}) where {T, R, N, M, RA}
+    copy!(dest.refs, view(parent(src).refs, src.indices))
+    Threads.atomic_add!(src.refcount, 1)
+    dest.pool = src.pool
+    dest.invpool = src.invpool
+    dest.refcount = src.refcount
+    return dest
+end
+
+function Base.copyto!(dest::PooledArray{T, R, N, RA}, doffs::Union{Signed, Unsigned},
+                 src::PooledArray{T, R, N, RA}, soffs::Union{Signed, Unsigned},
+                 n::Union{Signed, Unsigned}) where {T, R, N, RA}
+    n == 0 && return dest
+    n > 0 || Base._throw_argerror()
+    if soffs < 1 || doffs < 1 || soffs+n-1 > length(src) || doffs+n-1 > length(dest)
+        throw(BoundsError())
+    end
+
+    if length(dest.pool) == 0
+        @assert length(dest.invpool) == 0
+        Threads.atomic_add!(src.refcount, 1)
+        dest.pool = src.pool
+        dest.invpool = src.invpool
+        Threads.atomic_sub!(dest.refcount, 1)
```

Comment:
We need this atomic_sub because the actual `refcount` objects are shared between `PooledArray`s, right? So some other PooledArray might also have this same `refcount`?

Reply:
Yes that's the point. The refcount is shared just like the pool it refers to.

Actually a somewhat cleaner design would be to have a `Pool` object wrapping `pool`, `invpool` and `refcount` since these are tightly linked. But that would require a bit of refactoring.

Reply:
Yeah, I had the same thought; probably a bit of work for a little benefit, but it would help solidify that the three fields are a bundle together.

Reply:
I avoided it (but actually had exactly the same thought - also CategoricalArrays.jl does exactly this), because @nalimilan commented earlier (and I agree with it) that since we are post 1.0 release changing this kind of things should probably bump the version to 2.0. Although `pool` and `invpool` are documented, it is likely that legacy code (especially pre DataAPI.jl) referenced to it.

Reply:
To be clear I don't think internal fields are part of the 1.0 API contract. But in general it's better to avoid changing things too much if that's not really needed.

Reply:
This is what I understood.